### PR TITLE
Implement app manager

### PR DESCRIPTION
**Pull request by L0czek**
This PR adds the app-manager prototype with mocked keysealing and application launcher. Later as other components become ready I will revisit it. To test this PR I provide a python script that acts as a mocked host daemon (warden_daemon). It allows for testing provisioning, starting, stopping and rebooting the realm and installed applications.

 The instruction on how to run and test this PR is available [here](https://github.com/islet-project/realm-manager/tree/dev/m.szaknis/implement_app_manager/realm/mocked_host)

### CODE REVIEW

File: realm/app-manager/src/manager.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    async fn report_provision_success(&mut self) -> Result<()> {
+        self.conn
+            .send(Response::Success())
+            .await
+            .map_err(ManagerError::FramedJsonError)?;
+
+        Ok(())
+    }
```
Comment (MJJSoftware) [inline realm/app-manager/src/manager.rs:111]: There is also `Response::ProvisioningFinished` which in fact isn't used anywhere in the code, did you consider using this request here? You can also delete previously mentioned response and keep this part of the code unchanged (which might be even the best option).

File: realm/mocked_host/run.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            disk.table.partitions.add(data_part)
+            disk.commit()
+
+    def run_qemu(self):
+        self.qemu = subprocess.Popen(shlex.split(f"""
+ "../tools/qemu/build/qemu-system-aarch64" \
+        -machine virt \
+        -cpu cortex-a57 \
+        -nographic -smp 1 \
+        -kernel /home/m.szaknis/dev/realm-manager/realm/linux/arch/arm64/boot/Image \
```
Comment (MJJSoftware) [inline realm/mocked_host/run.py:58]: Could you transform this hard coded value into an input variable?

File: realm/mocked_host/run.py
Comment (MJJSoftware) [inline realm/mocked_host/run.py:1]: I've run this test script and unfortunately there is an error:
```
[app_manager::app] Mounting overlayfs
Error: UtilError(Fs(MountError(ENODEV)))
```
and the script crashes.

File: realm/mocked_host/run.py (hunk 2 above)
Comment (L0czek) [reply realm/mocked_host/run.py:58]: Uffff sorry, I didn't catch that.

File: realm/app-manager/src/util/os.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    unsafe {
+        nix::libc::sync();
+        nix::libc::reboot(op);
+    };
+
+    error!("Reboot has failed");
+
+    Err(OsError::RebootError(Errno::last()).into())
+}
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/util/os.rs:33]: It is just rebooting, the call to sync() is required to make sure that all kernel buffers are flushed.

File: realm/app-manager/src/util/os.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#[derive(Debug, Error)]
+pub enum OsError {
+    #[error("Reboot error")]
+    RebootError(#[source] Errno),
+}
+
+pub enum RebootAction {
+    Reboot,
+    Shutdown,
+}
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/util/os.rs:17]: Sure, we can go with "SystemPowerAction". I named it "RebootAction" as the syscall also takes an "REBOOT_OP param.

File: realm/app-manager/src/util/mod.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    CstringConvError(#[from] NulError),
+
+    #[error("Vector conversion error to CString")]
+    CstringFromVecConvError(#[from] FromVecWithNulError),
+
+    #[error("OS error")]
+    Os(#[from] OsError),
+}
+
+pub type Result<T> = std::result::Result<T, UtilsError>;
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/util/mod.rs:37]: Sure, can do

File: realm/app-manager/src/main.rs [hunk 6]
```diff
@@ -1,3 +1,65 @@
...
+    #[error("Utilities error")]
+    UtilError(#[from] UtilsError),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let args = Args::parse();
+    simple_logger::init_with_level(args.log_level).unwrap();
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/main.rs:52]: I think that expect is suffiecient as nobody will be able to handle this error as this is the main function.

File: realm/app-manager/src/app.rs [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+#[derive(Serialize, Deserialize)]
+pub struct ApplicationMetadata {
+    salt: Vec<u8>,
+}
+
+pub struct Application {
+    info: ApplicationInfo,
+    workdir: PathBuf,
+    devicemapper: Arc<DeviceMapper>,
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/app.rs:45]: The DM object is also kept inside the create devices as calling suspend() and resume() requires access to the device mapper object.

File: realm/app-manager/src/app.rs [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+    pub fn id(&self) -> &Uuid {
+        &self.info.id
+    }
+
+    fn get_handler(&mut self) -> Result<&mut (dyn ApplicationHandler + Send + Sync)> {
+        Ok(self
+            .handler
+            .as_mut()
+            .ok_or(ApplicationError::NotInstalled())?
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/app.rs:264]: Maybe, I will check it.

File: realm/mocked_host/run.py
Comment (L0czek) [reply realm/mocked_host/run.py:1]: Eh, maybe I didn't commit the last changes to the kernel config. It looks like the kernel image you have compiled doesn't have the support for OverlayFS.

File: realm/app-manager/src/manager.rs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        }
+    }
+
+    async fn perform_reboot(&mut self, action: RebootAction) -> ProtocolResult<Response> {
+        self.shutdown_all_apps().await;
+        match reboot(action) {
+            Ok(_) => unreachable!(), // Will never reach here
+            Err(e) => Err(ProtocolError::RebootActionFailed(format!("{:?}", e))),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/manager.rs:178]: Sure, I will fix it.

File: realm/app-manager/src/manager.rs (hunk 1 above)
Comment (L0czek) [reply realm/app-manager/src/manager.rs:111]: Yeah, you are right, I removed it.

File: realm/app-manager/src/app.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    keyring: KernelKeyring,
+    handler: Option<Box<dyn ApplicationHandler + Send + Sync>>,
+    devices: Vec<CryptDevice>,
+}
+
+impl Application {
+    pub fn new(info: ApplicationInfo, workdir: impl ToOwned<Owned = PathBuf>) -> Result<Self> {
+        Ok(Self {
+            info,
+            workdir: workdir.to_owned(),
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/app.rs:55]: done

File: realm/app-manager/src/util/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/util/mod.rs:37]: A second thought, I think that to be consistent every other such type definitions should be rewritten. You suggest to change only one module, I  think that it better to either change all of them or none to follow a established convetion.

File: realm/app-manager/src/manager.rs [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            }
+
+            Request::Reboot() => {
+                info!("Performing system reboot");
+                self.perform_reboot(RebootAction::Reboot).await
+            }
+        }
+    }
+
+    async fn handle_valid_request(&mut self, request: Request) -> Response {
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/manager.rs:245]: done

File: realm/app-manager/src/manager.rs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+                match app.try_wait().await {
+                    Ok(Some(status)) => Ok(Response::ApplicationExited(status.into_raw())),
+                    Ok(None) => Ok(Response::ApplicationIsRunning()),
+                    Err(MainError::LauncherError(LauncherError::HandlerError(
+                        ApplicationHandlerError::AppNotRunning(),
+                    ))) => Ok(Response::ApplicationNotStarted()),
+                    Err(e) => Err(ProtocolError::ApplicationWaitFailed(format!("{:?}", e))),
+                }
+            }
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/manager.rs:231]: How about "CheckStatus"?

File: realm/app-manager/src/manager.rs [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+        while let Some(result) = set.join_next().await {
+            let app = result.map_err(ManagerError::ProvisionJoinError)??;
+            let id = *app.id();
+            self.apps.insert(id, app);
+            info!("Finished installing {}", id);
+        }
+
+        info!("Provisioning finished");
+        self.report_provision_success().await?;
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/manager.rs:144]: done

File: realm/app-manager/src/manager.rs [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+        if self.config.autostartall {
+            for (id, app) in self.apps.iter_mut() {
+                info!("Starting {:?}", id);
+                app.start().await?;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/manager.rs:151]: done

File: realm/app-manager/src/app.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (L0czek) [reply realm/app-manager/src/app.rs]: done

File: realm/mocked_host/run.py (hunk 2 above)
Comment (L0czek) [reply realm/mocked_host/run.py]: done

File: realm/app-manager/src/util/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MJJSoftware) [reply realm/app-manager/src/util/mod.rs:37]: Sure I understand your point of view I am closing this conversation.

File: realm/app-manager/src/manager.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MJJSoftware) [reply realm/app-manager/src/manager.rs]: Nice idea

File: realm/mocked_host/run.py
Comment (MJJSoftware) [reply realm/mocked_host/run.py:1]: I've rebuilt the kernel and unfortunately I still encounter the same problem.

Comment (MJJSoftware) [reply realm/mocked_host/run.py:1]: I've deleted all files and compiled from scratch and then everything stared work.

Minor: Are there going to be similar backward compatible breaking changes in the future? If so, maybe we could put some information about cleaning previously built sources if something is not working fine?

Comment (L0czek) [reply realm/mocked_host/run.py:1]: Can't say, but the kernel config already has all required options enabled, so it's rather unlikely. However, I agree that we should put something inside the README to help with resolving such issues.

File: protocol/warden_realm/src/error.rs [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ProtocolError {
+    InvalidRequest(String),
+    ProvisionInfoNotReceived(),
+    ApplicationsAlreadyProvisioned(),
+    ApplicationNotFound(),
+    ApplicationLaunchFailed(String),
+    ApplicationStopFailed(String),
+    ApplicationKillFailed(String),
+    ApplicationWaitFailed(String),
```
[the remark this answers is not in the document]
Comment (L0czek) [reply protocol/warden_realm/src/error.rs]: I renamed it to ApplicationCheckStatusFailed.

File: realm/app-manager/src/app.rs [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    dirname, formatfs, mkdirp, mknod, mount, mount_overlayfs, read_to_string, stat, umount,
+    write_to_string, Filesystem,
+};
+use crate::util::serde::{json_dump, json_load};
+
+#[derive(Debug, Error)]
+pub enum ApplicationError {
+    #[error("Partition not found")]
+    PartitionNotFound(),
+}
```
Comment (MJJSoftware) [inline realm/app-manager/src/app.rs:30]: Minor: Maybe we could add information about partition (Uuid) that can't be found?

**Review by MJJSoftware — APPROVED** [review]

File: realm/app-manager/src/app.rs (hunk 16 above)
Comment (L0czek) [reply realm/app-manager/src/app.rs]: done

**Pull request merged**